File: AFMobile/AFAndroid/app/src/main/java/cz/cvut/fel/matyapav/afandroid/builders/FormBuilder.java
package cz.cvut.fel.matyapav.afandroid.builders;

import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.tomscz.afrest.commons.SupportedComponents;
import com.tomscz.afrest.layout.definitions.LayouDefinitions;
import com.tomscz.afrest.layout.definitions.LayoutOrientation;

import cz.cvut.fel.matyapav.afandroid.AFAndroid;
import cz.cvut.fel.matyapav.afandroid.components.types.AFComponent;
import cz.cvut.fel.matyapav.afandroid.components.parts.AFField;
import cz.cvut.fel.matyapav.afandroid.components.types.AFForm;

/**
 * Builds for from class definition
 * Created by Pavel on 26.12.2015.
 */
public class FormBuilder extends AFComponentBuilder<FormBuilder>{

    @Override
    public AFForm createComponent() throws Exception {
        initializeConnections();
        AFForm form = new AFForm(getActivity(), getConnectionPack(), getSkin());
        form.getModelResponse();
        //create form from response
        buildComponent(form);
        //fill it with data (if there are some)
        String data = form.getDataResponse();
        if(data != null) {
            form.insertData(form.getDataResponse());
        }
        AFAndroid.getInstance().addCreatedComponent(getComponentKeyName(), form);
        return form;
    }

    @Override
    protected View buildComponentView(AFComponent form) {
        LinearLayout formView = new LinearLayout(getActivity());
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        params.setMargins(getSkin().getComponentMarginLeft(), getSkin().getComponentMarginTop(),
                getSkin().getComponentMarginRight(), getSkin().getComponentMarginBottom());
        formView.setLayoutParams(params);
        //set form layout orientation
        if(form.getComponentInfo().getLayout().getLayoutOrientation().equals(LayoutOrientation.AXISX)){ //AXIS X
            formView.setOrientation(LinearLayout.VERTICAL);
        }else { //AXIS Y
            formView.setOrientation(LinearLayout.HORIZONTAL);
        }

        //set fields layout orientation
        int setOfFieldsOrientation;
        if(form.getComponentInfo().getLayout().getLayoutOrientation().equals(LayoutOrientation.AXISX)){ //AXIS X
            setOfFieldsOrientation = LinearLayout.HORIZONTAL;
        }else{ //AXIS Y
            setOfFieldsOrientation = LinearLayout.VERTICAL;
        }

        //determine layout
        int numberOfColumns;
        if(form.getComponentInfo().getLayout().getLayoutDefinition().equals(LayouDefinitions.TWOCOLUMNSLAYOUT)){
            numberOfColumns = 2;
        }else {
            numberOfColumns = 1;
        }

        int i = 0;
        LinearLayout setOfFields = null;;
        for (AFField field : form.getFields()) {
            if(!field.getFieldInfo().getVisible()){
                continue;
            }
            if (i % numberOfColumns == 0) {
                if(setOfFields != null) {
                    formView.addView(setOfFields);
                }
                setOfFields = new LinearLayout(getActivity());
                setOfFields.setOrientation(setOfFieldsOrientation);
                setOfFields.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT));
            }

            View fieldView = field.getCompleteView();
            fieldView.setLayoutParams(new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f / numberOfColumns));
            setOfFields.addView(fieldView);

            if(i == form.getVisibleFieldsCount() - 1){
                formView.addView(setOfFields);
            }

            i++;
        }

        return formView;
    }
}

File: AFMobile/AFWinPhone/AFWinPhone/builders/FormBuilder.cs
using System;
using System.Text;
using System.Threading.Tasks;
using AFWinPhone.components.types;
using Windows.UI.Xaml;
using AFWinPhone.enums;
using Windows.UI.Xaml.Controls;
using AFWinPhone.components.parts;

namespace AFWinPhone.builders
{
    /// <summary>
    /// This class if responsible for building form component, which should mainly hadle user input.
    /// </summary>
    public class FormBuilder : AFComponentBuilder<FormBuilder>
    {
        public override AFComponent createComponent()
        {
            this.initializeConnections();
            AFForm form = new AFForm(getConnectionPack(), getSkin());

            buildComponent(form);

            var dataTask = Task.Run(form.getDataResponse);
            dataTask.Wait();
            String data = dataTask.Result;
            if (data != null)
            {
                form.insertData(data);
            }

            AfWindowsPhone.getInstance().addCreatedComponent(this.getComponentKeyName(), form);
            return form;
        }

        protected override FrameworkElement buildComponentView(AFComponent form)
        {
            StackPanel formView = new StackPanel();
            formView.Margin = new Thickness(getSkin().getComponentMarginLeft(), getSkin().getComponentMarginTop(),
                getSkin().getComponentMarginRight(), getSkin().getComponentMarginBottom());
            //set form layout orientation
            if (form.getComponentInfo().getLayout().getLayoutOrientation().Equals(LayoutOrientation.AXISX))
            { //AXIS X
                formView.Orientation = Orientation.Vertical;
            }
            else { //AXIS Y
                formView.Orientation = Orientation.Horizontal;
            }

            //set fields layout orientation
            Orientation setOfFieldsOrientation;
            if (form.getComponentInfo().getLayout().getLayoutOrientation().Equals(LayoutOrientation.AXISX))
            { //AXIS X
                setOfFieldsOrientation = Orientation.Horizontal;
            }
            else { //AXIS Y
                setOfFieldsOrientation = Orientation.Vertical;
            }

            //determine layout
            int numberOfColumns = form.getComponentInfo().getLayout().getLayoutDefinition().getNumberOfColumns();

            int i = 0;
            StackPanel setOfFields = null; ;
            foreach (AFField field in form.getFields())
            {
                if (!field.getFieldInfo().isVisible())
                {
                    continue;
                }
                if (i % numberOfColumns == 0)
                {
                    if (setOfFields != null)
                    {
                        formView.Children.Add(setOfFields);
                    }
                    setOfFields = new StackPanel();
                    setOfFields.Orientation = setOfFieldsOrientation;
                }

                FrameworkElement fieldView = field.getCompleteView();
                setOfFields.Children.Add(fieldView);

                if (i == form.getVisibleFieldsCount() - 1)
                {
                    formView.Children.Add(setOfFields);
                }

                i++;
            }

            return formView;
        }
    }
}
